feat(devtools): implement iframe support

AleksanderBodurri opened the pull request:
Introduces Iframe support into Angular DevTools, see individual commits

JeanMeche commented on devtools/projects/shell-browser/src/app/app.component.ts:
```diff
@@ -23,21 +23,24 @@ import {ZoneAwareChromeMessageBus} from './zone-aware-chrome-message-bus';
 })
 export class AppComponent implements OnInit {
-  constructor(private _cd: ChangeDetectorRef) {}
+  environment = inject(ApplicationEnvironment);
```
nit: could be private also right ?

JeanMeche commented on devtools/projects/shell-browser/src/app/app.module.ts:
```diff
@@ -16,7 +17,7 @@ import {ChromeApplicationOperations} from './chrome-application-operations';
 @NgModule({
   declarations: [AppComponent],
-  imports: [BrowserAnimationsModule, DevToolsModule],
+  imports: [BrowserAnimationsModule, DevToolsModule, MatSelectModule],
```
What version of Material do we rely on ? In 17.1 all components are available as standalone.

JeanMeche commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
-  return +str + '' === str;
-};
+function registerPortTab(port: chrome.runtime.Port, frameId: string, tabId: string) {
+  ports[tabId] = ports[tabId] ?? {
```
```suggestion
  ports[tabId] ??= {
```

dgp1130 replied:
Question: Are we sure `??= { /* ... */ }` is the right operation here? If `ports[tabId]` already exists, this won't clear `devtools` or `content-script`. I wonder if we intended:

```typescript
ports[tabId] ??= {};
const tab = ports[tabId]!;
tab.devtools = null;
tab['content-script'] = null;
tab.contentScripts = { /* ... */ };

// ...
```

AleksanderBodurri replied:
Consider the case where the tab exists but we're registering a new frame. I think the name of this function is causing confusion here. Will rename it appropriately 🙏

AleksanderBodurri replied:
refactored this section, take a look and let me know what you think 🙏

JeanMeche commented on devtools/projects/shell-browser/src/app/content-script.ts:
```diff
@@ -29,8 +34,47 @@ export const main = () => {
+
+  detectAngularMessageBus.on('detectAngular', (detectionResult) => {
+    // only install backend once
+    if (backendInstalled) {
+      return;
+    }
+
+    if (detectionResult.isAngularDevTools !== true) {
+      return;
+    }
+
+    if (detectionResult.isAngular !== true) {
+      return;
+    }
+
+    if (detectionResult.isDebugMode !== true) {
+      return;
+    }
+
+    if (detectionResult.isSupportedAngularVersion !== true) {
+      return;
+    }
+
+    // Defensive check against non html page. Realistically this should never happen.
+    if (document.contentType !== 'text/html') {
+      return;
+    }
```
```suggestion
if (
  backendInstalled ||
  detectionResult.isAngularDevTools !== true ||
  detectionResult.isAngular !== true ||
  detectionResult.isDebugMode !== true ||
  detectionResult.isSupportedAngularVersion !== true ||
  // Defensive check against non html page. Realistically this should never happen.
  document.contentType !== 'text/html'
) {
  return;
}
```

And in case the formatter  doesn't keep 1 condition per line, you can add a trailing comment to force this layout !

dgp1130 replied:
Personally I prefer multiple `if` statements so I don't have to try to under the `||` combinations and each can get its own comments as necessary. I don't feel that strongly about it though.

AleksanderBodurri replied:
I prefer the multiple as well. 1 if statement with multiple unrelated criteria can get difficult to parse visually as combinations of `&&`, `||`, `===`, `!==` start getting used.

JeanMeche commented on devtools/projects/ng-devtools/src/lib/devtools.component.ts:
```diff
@@ -37,6 +38,8 @@ export class DevToolsComponent implements OnInit, OnDestroy {
   private readonly _firefoxStyleName = 'firefox_styles.css';
   private readonly _chromeStyleName = 'chrome_styles.css';
 
+  environment = inject(ApplicationEnvironment);
```
nit: Doesn't look it's used in the template, we should probably make it private

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
+
+      Object.keys(tab.contentScripts).forEach((key) => {
+        tab.contentScripts[key].enabled = false;
+      });
```
Style: Prefer native `for-of` to be more clear about the mutation here.

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
+      const tabId = message.args[1] as string;
+      const tab = ports[tabId];
+      if (tab?.contentScripts === undefined) {
+        return;
```
Question: Should this be treated as an error? Is it a problem if DevTools tries to connect but we don't have a content script to link it to?

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
+      devtoolsPort.postMessage({
+        topic: 'contentScriptConnected',
+        args: [contentScriptConnection.frameId, contentScriptConnection.port!.name]
+      });
```
Question: Should we do an early return here? Would we want the `backendRequest` message to continue on to DevTools?

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
-    doublePipe(portsTab.devtools, portsTab['content-script'], tab);
+  if (!port.sender || !port.sender.tab || port.sender.tab.id === undefined ||
+      port.sender.frameId === undefined) {
+    return;
```
Question: Do we expect ports to sometimes not have `tabId` or `frameId`? Should we throw an error / log something in this case?

AleksanderBodurri replied:
Good point, added a warning log for this case. In practice this case shouldn't happen

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
-        });
+      Object.entries(tab.contentScripts).forEach(([frameId, connection]) => {
+        connection.pipeSetup = false;
+      });
```
Style: Prefer `for-of` to be more clear about mutation.

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
+    tab.contentScripts[frameId] = tab.contentScripts[frameId] ?? {port: null, enabled: false};
+    tab.contentScripts[frameId].port = port;
+    tab.contentScripts[frameId].frameId = frameId;
+    tab.contentScripts[frameId].enabled = tab.contentScripts[frameId].enabled ?? false;
```
Suggestion: Use a temporary variable (and `??=`).

```typescript
tab.contentScripts[frameId] ??= {port: null, enabled: false};
const contentScript = tabs.contentScripts[frameId];
contentScript.port = port;
contentScript.frameId = frameId;
contentScript.enabled = enabled;
```

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
+    tab.contentScripts[frameId].frameId = frameId;
+    tab.contentScripts[frameId].enabled = tab.contentScripts[frameId].enabled ?? false;
+    port.onDisconnect.addListener(() => {
+      const port = tab.contentScripts[frameId].port;
```
Question: Are frame IDs ever reused? Is there a risk that a different content script exists at `tab.contentScripts[frameId]` during the `port.onDisconnect` listener vs what was present during `registerPortTab`?

Basically I'm wondering if this line should use the `contentScript` temporary variable I proposed above or if we specifically want to look up the frame ID each time?

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
+    port.onDisconnect.addListener(() => {
+      const port = tab.contentScripts[frameId].port;
+      if (!port) {
+        return;
```
Question: Do we want to run `delete tab.contentScripts[frameId];` or `delete ports[tabId];` if there is no `port`? Is this `return` a little too eager maybe?

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
+      tab.devtools!.postMessage(
+          {topic: 'contentScriptConnected', args: [frameId, connection.port.name]});
+      doublePipe(tab.devtools, connection);
+    });
```
Style: Prefer `for-of` for mutations.

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
-    return 'unsupported.html';
+
+  if (tab.contentScripts[frameId] && frameId !== 'devtools') {
+    tab.contentScripts[frameId].port!.onMessage.addListener((message) => {
```
Question: Does this listener ever get removed? Would it create a memory leak?

AleksanderBodurri replied:
This should get GC'd when we delete the reference to `tab.contentScripts[frameId]`

dgp1130 commented on devtools/src/demo-application-environment.ts:
```diff
@@ -11,7 +11,16 @@ import {ApplicationEnvironment, Environment} from 'ng-devtools';
 import {environment} from './environments/environment';
 
 export class DemoApplicationEnvironment extends ApplicationEnvironment {
+  multipleFramesEnabled: false = false;
```
Question: What's the motivation for disabling multiple frames? How do we use this environment?

AleksanderBodurri replied:
This was poorly named, renamed it to better reflect it's meaning

dgp1130 commented on devtools/projects/shell-browser/src/app/content-script.ts:
```diff
@@ -9,14 +9,19 @@
+  if ((window as any).__ANGULAR_DEVTOOLS_CONTENT_SCRIPT_INJECTED__) {
+    return;
+  }
+  (window as any).__ANGULAR_DEVTOOLS_CONTENT_SCRIPT_INJECTED__ = true;
```
Question: Is there a risk of running this content script on the same page multiple times? How does that happen?

On a related point, we could consider a `Symbol` over this property name if we're confident the content script itself won't be duplicated.

AleksanderBodurri replied:
Removed these entirely, no longer necessary

dgp1130 commented on devtools/projects/shell-browser/src/app/content-script.ts:
```diff
@@ -9,14 +9,19 @@
+let backendInstalled = false;
+
 export const main = () => {
+  if ((window as any).__ANGULAR_DEVTOOLS_CONTENT_SCRIPT_INJECTED__) {
```
Suggestion: Use `declare global` to define this type.

```typescript
declare global {
  var __ANGULAR_DEVTOOLS_CONTENT_SCRIPT_INJECTED__?: boolean;
}
```

dgp1130 commented on devtools/projects/shell-browser/src/app/content-script.ts:
```diff
@@ -9,14 +9,19 @@
 
+let backendInstalled = false;
+
 export const main = () => {
```
Suggestion: Can we drop this `export` now? Looks like it's just invoked immediately.

AleksanderBodurri replied:
Refactored to remove the iffe and include all the code at the top level of the script

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
Suggestion: Do we have any documentation on how the Chrome extension is architected and how messages are passed around?

I feel like I have a rough understanding of this but I'm sure there are gaps in my knowledge. It might be useful to have a markdown doc or Mermaid graph showing how messages move around the extension and the page. It would help a lot to better understand what `background` is actually trying to do here.

This is probably something we can follow up on in a separate PR, but something to consider as we make this system more complicated to support multiple frames.

AleksanderBodurri replied:
Good suggestion, will add to the devtools sync doc

dgp1130 commented on devtools/projects/shell-browser/src/manifest/manifest.chrome.json:
```diff
@@ -27,6 +27,7 @@
         "app/backend_bundle.js",
+        "app/content_script_bundle.js",
         "app/detect_angular_for_extension_icon_bundle.js",
         "devtools.html"
```
Question: Do these resources need to be web accessible? I feel like content scripts, and `devtools.html` shouldn't require this. I'm less clear on how backend bundle and detect Angular get loaded.

dgp1130 commented on devtools/projects/shell-browser/src/manifest/manifest.chrome.json:
```diff
@@ -52,7 +53,18 @@
+      "js": [
+        "app/content_script_bundle.js"
+      ],
+      "run_at": "document_idle",
```
Question: Would `document_idle` be a problem for users trying to inspect on page load? For example if the user breakpoints in `ngOnInit`. Would the extension still be usable in that context or would it be waiting to run the content script?

AleksanderBodurri replied:
This is true, but I do think this would also hold true for any of the alternatives for `run_at`

AleksanderBodurri replied:
The debugger would still block the main thread once ngOnInit triggers, preventing the backend and content script from communicating any messages to the background page.

I think there is a case to be made for using `document_start` though. Although it'll be a bit tricky to implement in a way where the `backend <-> content script <-> background page` communication is setup properly on a hard reload. Will add this to the sync doc but I think we're fine to defer this to later

dgp1130 replied:
Ah I see. Yeah it might be worth thinking through what features we can support in a breakpointed state. Benchmarking CD wouldn't be viable, but I feel like most of the components tab should be viable? DI conceptually could work, but might not be viable since the app will be stalled. Something we should think about and consider in the future, I agree we can defer it from this PR.

dgp1130 reviewed (commented):
Thanks for putting this together! Most of these comments are relatively minor.

I would love to see some automated tests here. There are a ton of `if { return; }` and null checks which comprehensive unit tests would help add a lot of confidence for. Is there anything we can do in that direction?

JoostK commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/devtools-tabs.component.html:
```diff
@@ -10,6 +10,13 @@
     </button>
   </div>
+
+  <select class="frame-selector" (change)="toggleFrame.emit($event.target.value)">
```
I suppose having multiple frames with Angular pages is somewhat rare, so I'd only show this when relevant (i.e. > 1 frame), to avoid cluttering the UI with irrelevant UI.

AleksanderBodurri replied:
This is a good point, although on the flip side since IFrames can load into a page at any time, the number of frames on page can jump from 1 to 1+ at any time as well. This would cause a shift of the layout assuming we were hiding this select until we have > 1 frames

JoostK replied:
Agreed that a layout shift should be avoided. I don't know about the current layout, I am thinking this could work if the select is shown as part of an existing row, where the presence of the dropdown wouldn't impact the layout.

dgp1130 replied:
+1 to avoiding layout shift, not totally sure on the UX we'd need to support that though. We could potentially render the drop down but disable it, though that might be confusing to users.

AleksanderBodurri replied:
This is how it currently looks

![Screenshot 2024-01-19 at 1 26 47 AM](https://github.com/angular/angular/assets/39253660/c1e852d8-338a-4b43-8524-8b9a1cd53cb4)


![Screenshot 2024-01-19 at 1 26 58 AM](https://github.com/angular/angular/assets/39253660/15512f28-f681-4187-80f8-9fb278c12204)

![Screenshot 2024-01-19 at 1 28 16 AM](https://github.com/angular/angular/assets/39253660/a451a244-f5e9-41bf-871a-184f894193c4)


![Screenshot 2024-01-19 at 1 33 17 AM](https://github.com/angular/angular/assets/39253660/eccbed22-467e-437f-9e84-f804f6d30c66)

![Screenshot 2024-01-19 at 1 33 29 AM](https://github.com/angular/angular/assets/39253660/01d08541-a624-4536-9d2d-d7b5cfbda0ec)

dgp1130 replied:
Looking at the screenshots, layout shift probably isn't that big a deal since it would only move the tabs (which the user won't click often), but honestly I don't think the frame selector is that obtrusive. The tab list is likely to remain relatively short anyways (not sure if we display a `>>` expander if it gets too long like native DevTools does).

I'd probably lean towards just leaving the selector there all the time. That makes it slightly more discoverable and avoids confusion when the selector isn't there ("Why can't I pick my iframe, the button is gone! Oh wait, I was on the wrong page and didn't have an iframe so the button was hidden...")

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/devtools.component.ts:
```diff
@@ -48,9 +51,60 @@ export class DevToolsComponent implements OnInit, OnDestroy {
+      }
+
+      if (this.environment.frames.length === 0) {
+        this.environment.selectedFrameId = null;
```
Brought this above the toggleFrame call and left a comment explaining it's purpose. There is a technically a case where it is possible to have no frames when an inspected page reloads, but realistically it should never happen since devtools has a listener for page navigations that also hard reloads DevTools.

AleksanderBodurri commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
+) {
+  if (contentScriptConnection.pipeSetup === true) {
+    return;
+  }
```
Refactored to remove the need for this variable entirely

AleksanderBodurri commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -36,152 +36,214 @@ browserAction.setIcon(
+interface ContentScriptConnection {
+  port: chrome.runtime.Port|null;
+  enabled: boolean;
+  frameId: string;
```
Refactored

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/application-environment/index.ts:
```diff
@@ -15,6 +15,16 @@ export interface Environment {
+  frameId: string;
+}
+
 export abstract class ApplicationEnvironment {
```
Refactored this. Take a look and let me know your thoughts 🙏

dgp1130 replied:
I like the `FrameManager` abstraction a lot, I think it makes things a lot cleaner here.

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/devtools.component.html:
```diff
@@ -1,58 +1,57 @@
       } @else {
-        <p class="text-message">
-          Angular Devtools only supports Angular versions 12 and above
+        <p class="text-message" matTooltip="A dev build is when the `optimization` flag is set to `true` in the angular.json config file.">
```
Oh my 😳  Good catch

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/devtools.component.ts:
```diff
@@ -9,12 +9,19 @@
 import {ThemeService} from './theme-service';
 
+enum AngularStatus {
+  UNKNOWN,
```
I think unknown makes more sense here. My feeling of this enum is

```ts
enum AngularStatus {
  /**
   * This page may have Angular but we don't know yet. We're still trying to detect it.
   */
  UNKNOWN,

  /**
   * We've given up on trying to detect Angular. We tried ${DETECT_ANGULAR_ATTEMPTS} times and
   * failed.
   */
  DOES_NOT_EXIST,

  /**
   * Angular was detected somewhere on the page.
   */
  EXISTS,
}
```

dgp1130 replied:
Ok, that seems reasonable. I'm fine with `UNKNOWN`.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-explorer.component.ts:
```diff
@@ -180,18 +184,60 @@ export class DirectiveExplorerComponent implements OnInit, OnDestroy {
+      return;
+    }
+
+    if (this._frameManager.urlHasMultipleFrames(selectedFrame.url)) {
```
Suggestion: I feel like we're repeating this logic in a few places. I wonder if we could move it into some kind of `FrameManager.getSelectedFrame()` function which would return the current frame or `undefined` if the top-level frame and warn / abort if multiple frames conflict? Basically a slightly more powerful version of the `selectedFrame` getter you already have.

It's a little awkward because `undefined` would be the top frame and also the case where there are multiple, so we might need a discriminated union there which adds complexity.

Related question: Do we need `undefined` to point at the top-level frame? Could we just use the frame ID in that case and simplify drop the `undefined` case in all this logic, or does that further complicate things?

AleksanderBodurri replied:
Not sure I'm following exactly but I replaced this with a more natural feeling `frameHasUniqueUrl(frame: Frame | null): boolean` that handles the null case by returning true

dgp1130 replied:
I meant something like:

```typescript
export class FrameManager {
  getSelectedFrame(): Frame | null {
    if (this._selectedFrameId === null) {
      return null;
    }

    const frame = this._frames.get(this._selectedFrameId) ?? null;
    if (frame && !this.frameHasUniqueUrl(frame)) {
      this._messageBus.emit('log', [ /* ... */ ]);
      return null;
    }
  }
}
```

This way `getSelectedFrame` automatically checks if the frame is unique so each usage doesn't need to do that.

I'm not sure if we ever need to distinguish between the `_selectedFrameId === null` case and the `!this.frameAsUniqueUrl(frame)` case, so hopefully `return null;` works in both of them without creating ambiguity. Otherwise we might need a [discriminated union](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) to disambiguate them.

dgp1130 commented on devtools/projects/shell-browser/src/app/background.ts:
```diff
@@ -37,135 +37,119 @@ browserAction.setIcon(
-    tab = port.sender.tab.id.toString();
-    name = 'content-script';
-  }
+(globalThis as any)['myPorts'] = ports;
```
Question: Is this intended or leftover from debugging something?

dgp1130 commented on devtools/projects/shell-browser/src/app/chrome-application-operations.ts:
```diff
@@ -11,32 +11,41 @@
+    chrome.devtools.inspectedWindow.eval(script);
+  } else {
+    chrome.devtools.inspectedWindow.eval(script, {frameURL});
+  }
```
Suggestion: Can we just do:

```typescript
chrome.devtools.inspectedWindow.eval(script, {frameUrl});
```

I think that would work in either case, even if `frameUrl` is `undefined`?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/frame_manager.ts:
```diff
@@ -0,0 +1,134 @@
+    });
+  }
+
+  isSelectedFrame(frameId: number): boolean {
```
Consider: Would this API be more uniform if the methods accepted / returned `Frame` objects consistently instead of operating on its fields like `frameId` and `frameUrl` where possible?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/frame_manager.ts:
```diff
@@ -0,0 +1,134 @@
+  }
+
+  inspectFrame(frameId: number) {
+    if (!this._environment.frameSelectorEnabled) {
```
Suggestion: I feel like this class shouldn't know or care whether or not the frame selector is enabled. Do the relevant methods not make sense in a single-frame environment where the selector is disabled? Should the frame selector be a UI-level concern handled by the components independently of this class?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/frame_manager.ts:
```diff
@@ -0,0 +1,134 @@
+    return Array.from(this._frames.values());
+  }
+
+  get tabId(): number | null {
```
Consider: Can we remove this? Do we use it anywhere?

I feel like if someone needs the inspected tab ID, they should be getting it themselves and not asking a frame manager for it.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/frame_manager.ts:
```diff
@@ -0,0 +1,134 @@
+
+  private _frameUrlToFrameIds = new Map<string, Set<number>>();
+
+  constructor(
```
Style: I generally recommend keeping constructors as pure as possible and only initialize state. Consider making the constructor private and move all the side effects of subscribing to the message bus over to a factory.

There's a general expectation that an expression like `new FrameManager()` won't create side effects, but `FrameManager.from` or `FrameManager.bind` more clearly may contain side effects.

dgp1130 reviewed (commented):
Thanks for working through all these comments! It's definitely looking a lot cleaner. Looking forward to the tests for this PR.

AleksanderBodurri committed:
refactor(devtools): implement iframe support for Angular DevTools' browser code

Modifies the messaging layer of devtools to allow for switching communication between frames on a page. When served as a browser extension.

Design:
- When a page renders, DevTools installs a content script onto it through it's manifest file. The all_frames option is used here to install this script onto every frame in a page.
- When Angular is detected, the content script will install a backend script into it's frame.
- Each content script / backend script pairing is kept track of in the background script. This pairing represents an angular devtools context in a particular frame.
- Angular DevTools is able to ask the background script to list each frame that has been registered on a page.
- Angular Devtools is able to ask the background script to "enable" the connection on a particular frame. This enables the messaging between the content script <-> background script <-> devtools page

Limitations:
- The `inspectedWindow.eval` API is only able to target frames by frameURL. This means some features that integrate with Chrome DevTools like inspect element and open source will not be available when inspecting frames that do not have a unique url on the page.

AleksanderBodurri committed:
refactor(devtools): implement multiframe support in devtools page

In the Angular DevTools Chrome DevTools page:

- Angular DevTools is able to ask the background script to list each frame that has been registered on a page.
- Angular Devtools is able to ask the background script to "enable" the connection on a particular frame. This enables the messaging between the content script <-> background script <-> devtools page
- Implements detection of non unique urls on the inspected page

Limitations:
- The `inspectedWindow.eval` API is only able to target frames by frameURL. This means some features that integrate with Chrome DevTools like inspect element and open source will not be available when inspecting frames that do not have a unique url on the page.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/frame_manager_spec.ts:
```diff
@@ -0,0 +1,203 @@
+    (frameManager as any)._inspectedWindowTabId = tabId;
+  });
+
+  it('initialize should setup message bus subscriptions', () => {
```
Suggestion: I wouldn't bother testing that callbacks have been added if the tests below are already testing the effects of those callbacks.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/frame_manager_spec.ts:
```diff
@@ -0,0 +1,203 @@
+    const topLevelFrame = getFrameFromFrameManager(topLevelFrameId);
+    const otherFrame = getFrameFromFrameManager(otherFrameId);
+    expect(frameManager.isSelectedFrame(topLevelFrame)).toBe(true);
+    expect(frameManager.isSelectedFrame(otherFrame)).toBe(false);
```
Suggestion: Move the "non-selected frame returns `false`" case into its own test or update this tests description to be more clear that's included in this test's scope.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/frame_manager_spec.ts:
```diff
@@ -0,0 +1,203 @@
+    expect(frameManager.selectedFrame?.id).toBe(topLevelFrameId);
+  });
+
+  it('inspectFrame should not set selected frame when frame does not exist', () => {
```
Consider: Might be good to throw an error or emit a warning in this case? Seems like an easy mistake to make which has no effect.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/frame_manager_spec.ts:
```diff
@@ -0,0 +1,203 @@
+
+  it('frameHasUniqueUrl should return true when only one frame has a given url', () => {
+    frameManager.initialize();
+    contentScriptConnected(topLevelFrameId, 'name', 'https://angular.dev/');
```
Suggestion: Connect a second frame with a different URL. Otherwise this logic could be broken but the test case still passes just because there's only one frame.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/frame_manager_spec.ts:
```diff
@@ -0,0 +1,203 @@
+  it('frameHasUniqueUrl should return false when frame is null', () => {
+    frameManager.initialize();
+    expect(frameManager.frameHasUniqueUrl(null)).toBe(false);
+  });
```
Suggestion: Should we also test `frameHasUniqueUrl` for the case with hash fragments? That seems like a particularly nuanced case which would be worth including here.

dgp1130 reviewed (approved):
Thanks for putting in the effort on these tests! I think they'll help a lot here as this code evolves over time.

I added a bunch of minor suggestions, but nothing is particularly concerning to me or necessarily blocking this PR. It's a little hard to tell how code moved around, so please ignore any comments which are actually about pre-existing code. I'm not trying to re-litigate existing code, it's just sometimes hard to tell where everything falls. Feel free to ignore or push back on any comments you don't agree with or don't want to bother with right now.

AleksanderBodurri commented on devtools/projects/shell-browser/src/app/tab_manager_spec.ts:
```diff
@@ -0,0 +1,324 @@
+
+import {DevToolsConnection, TabManager, Tabs} from './tab_manager';
+
+interface MockSender {
```
Sender is a chrome runtime type, not provided by us. I don't see too much value in mocking the sender to spec perfectly, we don't use any fields other than id on the production side of this code. If we start using sender more then we can revisit this and add expand the mock type

AleksanderBodurri commented on devtools/projects/shell-browser/src/app/tab_manager_spec.ts:
```diff
@@ -0,0 +1,324 @@
+const TEST_MESSAGE_ONE = {topic: 'test', args: ['test1']};
+const TEST_MESSAGE_TWO = {topic: 'test', args: ['test2']};
+
+class MockPort {
```
Same as above https://github.com/angular/angular/pull/53934#discussion_r1484069098

AleksanderBodurri commented on devtools/projects/shell-browser/src/app/tab_manager_spec.ts:
```diff
@@ -0,0 +1,324 @@
+
+  function emitDisconnectToPort(port: MockPort) {
+    port.onDisconnectListeners.forEach((listener) => listener());
+  }
```
I considered this too. These are technically events that happen **to** a port, not necessarily something that a port does itself

AleksanderBodurri commented on devtools/projects/shell-browser/src/app/tab_manager_spec.ts:
```diff
@@ -0,0 +1,324 @@
+
+    it('should set tab.devtools to null when the devtoolsPort disconnects', () => {
+      emitDisconnectToPort(devtoolsPort);
+      expect(tab.devtools).toBeNull();
```
The devtools port disconnecting doesn't necessarily mean that we need to clear `tabs`. Consider the case where you have Angular Devtools open and then close chrome Devtools. You should be able to reopen chrome devtools without refreshing your page and be able to access devtools again. If we deleted the content script data from tabs when devtools disconnects, we'd be unable to reconnect to the page when devtools opens again.

dgp1130 replied:
Sounds reasonable, thanks for clarifying.

dgp1130 reviewed (approved):
Thanks for working through all the comments!

AndrewKushnir commented:
FYI, adding the "cleanup" label, since commit messages require an update (`feat` -> `refactor`).

AndrewKushnir commented:
This PR was merged into the repository by commit ebcdc8dc96e971f68e30cfe4acda5eba87417d77.
